src/vm/methodtablebuilder.cpp

```diff
@@ -3828,6 +3829,20 @@ VOID    MethodTableBuilder::InitializeFieldDescs(FieldDesc *pFieldDescList,
             }
         }
 
+        // Determine if a field is a fixed buffer.
+        {
+            HRESULT hr;
+
+            hr = pInternalImport->GetCustomAttributeByName(bmtMetaData->pFields[i],
```

Comment:
Just as a note. It is somewhat unfortunate that this detection is limited to `FixedBufferAttribute`.

The C# compiler currently only supports a limited set of types for its `fixed-sized buffers` and it currently results in an error (`error CS1716: Do not use 'System.Runtime.CompilerServices.FixedBuffer' attribute. Use the 'fixed' field modifier instead.`) if you try to use it for custom types.

This means that, in some scenarios, it may be impossible to create a struct with the appropriate layout. But, perhaps, this is something that can be fixed on the compiler end of things....

Reply:
At worst, the user can create a structure with each of the elements of the array as separate fields. The code in here for dealing with `FixedBufferAttribute` is more of a workaround for the fact that Roslyn only emits metadata for one field in the generated structure for a fixed buffer and that the attribute is placed on the field and not the type.

Reply:
Also as a quick note, for compatibility reasons, the runtime will likely never support correctly marshalling fixed buffers of non-blittable types.

Reply:
The HFA classifier for ARM/ARM64 deals with this case by assuming that all fields in the struct with explicitly specified size are of the same type, without checking the fixed attribute: https://github.com/dotnet/coreclr/blob/master/src/vm/class.cpp#L1316 . Can we do the same here?

Also, do we need to make matching fixes in the HFA classifier as well to make the explicit layout work better across all platforms (ARM/ARM64)? It has the same block for explicit layout: https://github.com/dotnet/coreclr/blob/master/src/vm/class.cpp#L1246

Reply:
> The HFA classifier for ARM/ARM64 deals with this case by assuming that all fields in the struct with explicitly specified size are of the same type, without checking the fixed attribute: https://github.com/dotnet/coreclr/blob/master/src/vm/class.cpp#L1316 . Can we do the same here?

We can't do the same logic. HFAs are unique in that they always have fields of the same type, and by the time the code gets to that line, that check has already been completed. We could probably do a check of "Sequential struct has one field of primitive value type, with an explicit size that is a multiple of that value type" to determine if a type is maybe a fixed buffer. That might catch some false-positives, but those false positives would be really weird ways of writing structures anyway.

> Also, do we need to make matching fixes in the HFA classifier as well to make the explicit layout work better across all platforms (ARM/ARM64)? It has the same block for explicit layout: https://github.com/dotnet/coreclr/blob/master/src/vm/class.cpp#L1246

Possibly? I haven't looked at the ARM/ARM64 calling convention docs. A quick fix for most explicit layout HFAs likely would be bailing out of there are any overlapping or unaligned fields. I'll see if I can find more information on how the HFA stuff works.

Reply:
> We could probably do a check of "Sequential struct has one field of primitive value type, with an explicit size that is a multiple of that value type" to determine

Right, this is what I meant. It is roughly what the HFA classifier does on ARM/ARM64.

Reply:
Ok I'll try that out and see if it works.